### PR TITLE
Add some more hooks

### PR DESCRIPTION
**Pull request by spencerhance**
Includes hooks for ILB and for https://github.com/kubernetes/ingress-gce/pull/781

### CODE REVIEW

File: pkg/cloud/mock/mock.go [hunk 1]
```diff
@@ -457,6 +457,26 @@ func UpdateRegionBackendServiceHook(ctx context.Context, key *meta.Key, obj *ga.
 	return nil
 }
 
+// UpdateRegionBackendServiceHook defines the hook for updating a Region
+// BackendsService. It replaces the object with the same key in the mock with
+// the updated object.
+func UpdateAlphaRegionBackendServiceHook(ctx context.Context, key *meta.Key, obj *ga.BackendService, m *cloud.MockAlphaRegionBackendServices) error {
+	_, err := m.Get(ctx, key)
+	if err != nil {
```
Comment (bowei) [inline pkg/cloud/mock/mock.go]: Through the error unless it is actually a NotFound errors

File: pkg/cloud/mock/mock.go [hunk 2]
```diff
@@ -552,6 +572,25 @@ func UpdateAlphaURLMapHook(ctx context.Context, key *meta.Key, obj *alpha.UrlMap
 	return nil
 }
 
+// UpdateAlphaURLMapHook defines the hook for updating an alpha UrlMap.
+// It replaces the object with the same key in the mock with the updated object.
+func UpdateAlphaRegionURLMapHook(ctx context.Context, key *meta.Key, obj *alpha.UrlMap, m *cloud.MockAlphaRegionUrlMaps) error {
+	_, err := m.Get(ctx, key)
+	if err != nil {
```
Comment (bowei) [inline pkg/cloud/mock/mock.go]: see comment above

File: pkg/cloud/mock/mock.go [hunk 3]
```diff
@@ -566,6 +605,34 @@ func SetTargetGlobalForwardingRuleHook(ctx context.Context, key *meta.Key, obj *
 	return nil
 }
 
+// SetTargetForwardingRuleHook defines the hook for setting the target proxy for a ForwardingRule.
+func SetTargetForwardingRuleHook(ctx context.Context, key *meta.Key, obj *ga.TargetReference, m *cloud.MockForwardingRules) error {
+	fw, err := m.Get(ctx, key)
+	if err != nil {
+		return &googleapi.Error{
```
Comment (bowei) [inline pkg/cloud/mock/mock.go]: see above

File: pkg/cloud/mock/mock.go [hunk 4]
```diff
@@ -594,6 +661,34 @@ func SetURLMapTargetHTTPProxyHook(ctx context.Context, key *meta.Key, ref *ga.Ur
 	return nil
 }
 
+// SetURLMapTargetHTTPProxyHook defines the hook for setting the url map for a TargetHttpProxy.
+func SetURLMapTargetHTTPSProxyHook(ctx context.Context, key *meta.Key, ref *ga.UrlMapReference, m *cloud.MockTargetHttpsProxies) error {
+	tp, err := m.Get(ctx, key)
+	if err != nil {
```
Comment (bowei) [inline pkg/cloud/mock/mock.go]: see above

File: pkg/cloud/mock/mock.go [hunk 5]
```diff
@@ -594,6 +661,34 @@ func SetURLMapTargetHTTPProxyHook(ctx context.Context, key *meta.Key, ref *ga.Ur
...
+	}
+
+	tp.UrlMap = ref.UrlMap
+	return nil
+}
+
+// SetURLMapTargetHTTPProxyHook defines the hook for setting the url map for a TargetHttpProxy.
+func SetURLMapAlphaRegionTargetHTTPSProxyHook(ctx context.Context, key *meta.Key, ref *alpha.UrlMapReference, m *cloud.MockAlphaRegionTargetHttpsProxies) error {
+	tp, err := m.Get(ctx, key)
+	if err != nil {
```
Comment (bowei) [inline pkg/cloud/mock/mock.go]: see above

File: pkg/cloud/mock/mock.go (hunk 1 above)
Comment (spencerhance) [reply pkg/cloud/mock/mock.go]: Okay I updated the error handling, not sure if that's what you had in mind

File: pkg/cloud/mock/mock.go (hunk 2 above)
Comment (spencerhance) [reply pkg/cloud/mock/mock.go]: done

File: pkg/cloud/mock/mock.go (hunk 3 above)
Comment (spencerhance) [reply pkg/cloud/mock/mock.go]: done

File: pkg/cloud/mock/mock.go (hunk 4 above)
Comment (spencerhance) [reply pkg/cloud/mock/mock.go]: done

File: pkg/cloud/mock/mock.go (hunk 5 above)
Comment (spencerhance) [reply pkg/cloud/mock/mock.go]: done

**Comment by bowei** [comment]
probably just pass err unchanged back to caller? It's up to the mock (and users of the mock) to use the right type of error

**Comment by spencerhance** [comment]
@bowei Makes sense to me - fixed in latest push.

File: pkg/cloud/mock/mock.go (hunk 1 above)
Comment (bowei) [reply pkg/cloud/mock/mock.go]: you need to fix this as well?

Comment (spencerhance) [reply pkg/cloud/mock/mock.go]: Good catch.  Since i'm making this change anyway I can also just update the other errors.

Comment (spencerhance) [reply pkg/cloud/mock/mock.go]: Done

**Pull request merged**